### DISCUSSION TITLE
LYWSD03MMC in the freezer

### DISCUSSION THREAD

**Discussion by evlo** [pvvx/ATC_MiThermometer#377] [Q&A] [answered]
I did put one LYWSD03MMC into the freezer and it stopped measuring at 3.55C and does not go lower, temperature is definitely bellow zero, batter voltage doped as expected. I remember previous winter some of these I have outside measured negative temperatures.

I do understand it is quite sudden temperature and voltage difference - but what is going on? why it stopped reporting at 3.55C? Will it go lower after some time?

![4nJvrWm](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/80fd283e-1488-4d4a-b59b-1d49f93ec0c1)

I can still connect to it using https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html and there i get -0.01C
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/5e3cab03-28c7-44af-a2af-09acb81aa0c0)

and now even
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/dc097064-0bdc-45b2-84b5-37e64c5edbe6)

Update:
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/c67a42e8-c35d-4215-b40d-9beecb62ab50)

**Comment by MacSass** [comment]
As far as I know those are intnded for indoor only and thus not rated for below 0 unfortunately. 

Measure range 0°C-60°C

Temperature display resolution0.1°C

Humidity range 0%99%RH

Humidity display resolution1 %RH

I had seen a previous discussion on this with the same conclusion.

**Comment by evlo** [comment]
It does measure bellow zero very well, I used it outside over the last winter and had very correct data.
STH40 is -40°C to +125°C, TLSR8251F512ET24 -40°C to +85°C

Weirdness continues
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/91baa942-e687-4174-aeef-15a2853ab83a)

Now it shows -16 and directly on the device -16 too, but why did it take so log to show in Ha?
It took about 30 minuts to update, but in that time it did update battery voltage in advertisement

Don't get me wrong, this is still better than any out of box commercial sensor I tried so far, but I'm quite curious why it it did not update (now it updates again).

I also tested that if advertisement is not received HA does NOT repeat old data and just reports nothing (correct behavior)

Now everything seems to work after a while, just reported +3.55 and then jumped to -15.99 after about 30 minutes

Another update, now:
Vbat: 2272 mV , Temp: -18.42°C, Humi: 50.69%, ID: 30560, Reed switch counter: 0, flg: 0x0E:r0/t1
and reporting -18.13 °C for about 10 minutes.

So I  move one of my esphome BT gateways near the freezer and it still reports -18.13 in HA while in TelinkMiFlasher  it show -18.41°C

Maybe it is just bad reception and for some reason it repeats last seen data as device is online or something?

**Comment by evlo** [comment]
weird that it is getting stuck like that
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/984c7ffd-c7d7-4899-a6f1-2dadae40b52f)

other devices that are same hw and fw do not have these dropouts
![image](https://github.com/pvvx/ATC_MiThermometer/assets/8813196/b679f199-7791-447b-a835-b3abe9f92e15)


maybe it is secret battery saving measure?

and to add to confusion display go stuck at -14 (thought similar thing with stuck display happened when i let on in the sun) and battery symbol was showing. 

Thought this one i used does have same 2032 in it for two years now, so I will buy more sensor and do a test with fresh batteries.
Maybe related https://github.com/pvvx/ATC_MiThermometer/issues/265#issuecomment-1431495494

**Reply by pvvx** [reply]
Regular CR2032 do not work in cold.
The typical maximum is -32C for a new expensive CR2032 battery. The voltage drops below 2.0 V.

**Comment by pvvx** [comment]
https://github.com/pvvx/BLE_THSensor

2 AAA battery

Work in the refrigerator freezer (BTHome v2, BLE LongRange for stable long-distance communication through the iron walls of the refrigerator):
![image](https://github.com/pvvx/ATC_MiThermometer/assets/12629515/0b1daac3-2cc9-452f-a593-26d3e005a4e9)

**Comment by scottcopus** [comment]
I currently just have a single LYWSD03MMC (HW B1.5 I think) for now but looking to buy more.  The hope is to put a few of them in my fridge/freezers.  BTW, if anyone has any good US sources for these let me know.

Does anyone know the _max_ battery supply voltage the LYWSD03MMC can accept?  I'm looking for a way to "mount" a square-ish rechargeable battery on the back of this and maybe rig a battery connector plug thru the side of it.  I've also got a lot of NiMH AA & AAA's and I guess I could always use a two-battery holder.  But is the minimum temperature limits for both NiMH and Lithium-ion batteries -20C?  Which is about what most freezers are running?  😱❄️

**Reply by DavidStacer** [reply]
I modified one to take a CR2450.  [https://github.com/pvvx/ATC_MiThermometer/discussions/396](https://github.com/pvvx/ATC_MiThermometer/discussions/396)
Just simple tools required.

**Reply by pvvx** [reply]
The maximum supply voltage for the chip is limited to 3.6V. When higher voltage is applied for a short period of time, the chip degrades. Subsequently, it will continue to work, but with increased consumption.

Such experiments were carried out repeatedly. After connecting a voltage above 4V, all chips degraded. When the voltage dropped to the normal level of 2..3.3V, such chips began to consume significantly more, but continued to work.

**Reply by scottcopus** [reply]
Thanks @DavidStacer. I actually saw your mod post. Thanks for sharing. I'll comment over there with a couple related questions soon.

Thanks @pvvx. Would you say the entire LYWSD03MMC thermometer (at the battery terminal) should be limited to 3.6V, or reduced further to prevent damage to other components like the sensor? If using a 3.7V lithium-ion or a 3-cell NiMH holder (1.2V x 3 = 3.6V), would a diode-based voltage drop work, or would you suggest sticking with an LDO like an MCP17xx or something else?  Just curious what you think.  For my cold freezer installations though, I might stick with regular alkaline 2 x 1.5v AA/AAA and see how the batteries fair after some weeks/months.

**Reply by pvvx** [reply]
Sometimes one diode is enough. Or two in series connection. But the best option is LDO.
MCP17xx is not the best in terms of its own consumption...

**Reply by pvvx** [reply]
> For my cold freezer installations though, I might stick with regular alkaline 2 x 1.5v AA/AAA and see how the batteries fair after some weeks/months.

https://github.com/pvvx/ATC_MiThermometer/discussions/377#discussioncomment-7698778

**Comment by scottcopus** [comment]
On my LYWSD03MMC I've got a reed switch connected on it.  But for some reason HA doesn't always catch when the reed switches on/off or can be delayed for several seconds.  My intent was to catch when my doors (fridge/freeze/any door) opens and closes.  Does anyone know what that is?  Are reed switch events always fired off thru BT instantly?  If not can be it configured to?

FYI, I'm using ESPHome bluetooth proxy on a ESP32 (TinyPico). 

BT firmware log:
4:46:24 PM: Hardware Version: LYWSD03MMC B1.5, Software Version: 4.3, Sensor: SHTC3 (SHTV3)
4:46:24 PM: Custom config HEX string: 55171000002804a931318ab4

**Reply by pvvx** [reply]
The message about the activation of the reed switch is transmitted 3 times with an interval of 50 ms.

On the receiving side, the first message about the activation of the reed switch is processed 5..7 ms after the change in the contact state.

On ESP devices running both Wi-Fi and BLE, there are periods of more than 1 second when they are not receiving anything. This is due to working with WiFi.

**Comment by pvvx** [comment, marked as answer]
> FYI, I'm using ESPHome bluetooth proxy on a ESP32 (TinyPico).

Reason for using ESP.
ESP misses a lot of receptions from BLE devices:

* Overloaded software. Low processing speed.
* Most of the time the antenna works on Wi-Fi.

ESP is not recommended for use in BLE.
Also, the ESP software processes receiving packets very slowly on a BLE connection. The battery device has to duplicate the transmission of each packet many times, which consumes a lot of battery power. Measurements show a fifty-fold increase in battery power consumption when connected to ESP.

**Reply by scottcopus** [reply]
@pvvx I'm not the OP, only a late commenter on this thread who happens to be using an ESPHome BT gateway too.  But if OP (@evlo) also happens to be using an ESP32 too, then your answer does help explain that could be the source of some of the issues.  👍

What are some good and low cost Bluetooth to Wifi gateways compatible with HA/ESPHome?  Are there dedicated BT modules that can relay BT advertisements/connection data to an ESP via serial/SPI/i2c/etc.... maybe Nordic nRF chips?  Since I don't have ethernet wired everywhere, I'm limited to wireless/wifi/RF.  😐

**Reply by pvvx** [reply]
The ESP program is created without using a FIFO to receive BLE packets and send them over Wi-Fi channels. After receiving the data, the program passes it to MQTT and waits for confirmation.
WiFi transmission has a step time of 102.4 ms during the WiFi Beacon signal period.
On the airwaves, if you have several BLE devices, their beacon transmissions are hundreds per second.
BLE advertising is especially common on smartphones, other BT adapters, TVs and Apple devices. Their typical transmission period for BT advertising is 50 ms.
As a result, the BLE receiver receives hundreds of BLE advertisements per second. Some of them have collisions. The receiver, in addition to having to process each packet, must also have time to adjust the automatic reception level to each packet and the noise level. And also switch through 3 channels, accompanying each advertising event.

Among inexpensive USB-BT adapters, the CSR8510A10 has good performance. The rest is about ten times worse.
The CSR8510A10 provides more or less stable reception of up to 50 BLE devices in Home Assistant.
RTL8761bu - up to 5.

* More or less stable reception - this means few missing points on the graphs in Home Assistant. This is not a limitation on the number of BLE devices supported.

The BT adapter built into the RPI is not worth considering at all.

No BT WiFi gateways can be normal unless they have separate Wi-Fi and Bluetooth chips and antennas separated by half a meter.
With Zigbee the situation is even tougher.

**Reply by pvvx** [reply]
Ultimately, it is desirable that BLE advertising packets be processed in less than 3 ms.
With direct programming, such as Arduino ESP with MQTT, or software for esp32 from Espressif, this cannot be achieved.
This is designed for gaming or learning, not for work devices...
The performance of an ESP32 with two cores, when the program does not have enough cache size from Flash, is at the level of 10..16 MHz Cortex M0. This situation always occurs due to the bloated size of the software.

**Reply by pvvx** [reply]
>  via serial/SPI/i2c/etc.... maybe Nordic nRF chips?

* "Nordic nRF" does not produce BT adapters.
* Linux does not yet fully support Bluetooth version 4.2, released in 2016. Support for Bluetooth version 5.0 and above is only available for Android

**Reply by scottcopus** [reply]
Thanks. Some really good info and thanks for sharing! 

Is it possible to tweak your custom firmware settings to make it work better with some of the known issues with the ESP BT+Wifi missing packets?  Even if ESPHome had a way to only filter for certain devices, would that even help or is the ESP hardware still a limiting factor?

What's the most economical method to get *_distant_* BT data reliably into HA, better than what an ESP BT gateway can do, even if using a dedicated USB-BT adapter like the CSR8510A10 is what works the best (which would need a USB host)?

Are you aware of any dedicated BT to wifi gateways, either with MQTT or having direct HA/BThome compatibility?

**Reply by pvvx** [reply]
I don't see the point in degrading the firmware to support the worst Bluetooth receivers.
This greatly limits all options, increases battery consumption and significantly increases the time it takes to read data from the device.
Why should anyone sacrifice functionality for Espressif's mediocrity?

**Reply by pvvx** [reply]
> What's the most economical method to get distant BT data reliably into HA, better than what an ESP BT gateway can do, even if using a dedicated USB-BT adapter like the CSR8510A10 is what works the best (which would need a USB host)?

Any BT sniffer is suitable for receiving BLE advertising.

**Reply by scottcopus** [reply]
Sorry for the confusion. I wasn't asking for a firmware code change. I was just asking if it was possible to tweak any *settings* that could improve when using the lacking ESP devices. That's why I also asked about switching to something else other then Espressif devices. ;)

What's your setup (or what would it be) for reliably receiving all BLE advertising packets from distant sensors and getting them into HA? 

When you say BT sniffer, are you referring to 'software'?  For example, is there an Android BT sniffer app that can relay directly to HA? If so an Android phone could be a distant BT gateway? Forgive my lack of knowledge in this area.

**Reply by pvvx** [reply]
There are firmwares for [nRF with sniffer ](https://www.nordicsemi.com/Products/Development-tools/nrf-sniffer-for-bluetooth-le)functions.
For other chips too.
Even this option accepts more packets than ESP - https://github.com/pvvx/AdScanerTrg

For Android you can directly in explorer and js - https://pvvx.github.io/ATC_MiThermometer/Advertising.html

Android has 'nRFconnect' and many other programs that work with BLE advertising in Bluetooth 5.0+ format (PHY 1M/2M/Coded: 250/125k - LE Long Range). They are written using the standard Android API. This is not possible on Linux and Windows.

[Bluetooth Web API](https://googlechrome.github.io/samples/web-bluetooth/), Bluez, [Bleak](https://github.com/hbldh/bleak/issues/1225), [Home Assistant](https://github.com/home-assistant/core/issues/88277), ... is limited (BT 4.2) due to Linux functionality (kernel and drivers not developing since 2016).

**Reply by pvvx** [reply]
> For example, is there an Android BT sniffer app that can relay directly to HA? If so an Android phone could be a distant BT gateway? 

These are commercial projects and devices. I am not their distributor. Let them make money from this disgrace in Linux.
When Linux is fixed, perhaps I will propose an open source project.

**Reply by evlo** [reply]
Kinda different topic, but I think reed switch might in the end be same thing as button press.

In my experience ESP32 (or ESP32-s3) is ok in connected mode (example here https://github.com/wifwucite/esphome-ble-controller ).
Even if it did not get 100% of the send messages (it does in my environment), there is under 100ms delay so people would just press the button again.
This is while both sender and receiver are esp32-s3 devices connected to the wifi AT THE SAME TIME too.

I'm still not 100% sure I fully understand difference between bluetooth connected and advertisement mode.

I do have 2 esp-s3 esphome gateways that do nothing else than being a gateway, and 4 that are on devices that do primarily other stuff.

For the advertised button press I still do not get quite a few in home assistant. 

I'm thinking in my target usage, when I would like to advertise button press and react on it (basically event driven wireless "bus") [so all the advertisers just advertise the press and all the receivers decide whether to react to it or not] there might even be issue that as there is lot of sensors broadcasting measurements, button press advertisement packets just collide with other stuff that is advertised on BLE at the same time. So the result is the same as when two people start talking on same VHF frequency on walkie talkie at the same time. But maybe ble is not affected by this?

My experience with other stuff like zigbee or z-wave battery powered devices is worse in the end result (pairing, coordinator, mesh vs gateways, no real control over he devices). I'm now thinking maybe rf433 for the battery operated buttons might be a way to go. [so far I found out that rf433 buttons I tested need to be held for exact time otherwise they do not send the whole message 🤦 ]

For the thermo hygro meters I don't care at all about few missed advertisements, for the buttons it makes the thing, for me, in the end, unusable.

Is there a way to use something like RTL8761 as a esphome ble gateway? (it works connected to the server running HA directly, but that is different thing). 
I do not even have HA server in the same county in one case :)

**Reply by pvvx** [reply]
> But maybe ble is not affected by this?

This is called a collision. For this reason, advertising messages in BLE are repeatedly duplicated.

It is impossible to confuse who the message is from. The message contains the MAC address of the device.

The WiFi specification has Bluetooth device detection to reduce collisions. But not in ESP.

**Comment by evlo** [comment]
> > For example, is there an Android BT sniffer app that can relay directly to HA? If so an Android phone could be a distant BT gateway?
> 
> These are commercial projects and devices. I am not their distributor. Let them make money from this disgrace in Linux. When Linux is fixed, perhaps I will propose an open source project.

There are? I'm not able to find anything, but having (old) phone act as HA ble gateway for advertisement would solve the processing power issue easily.

Maybe again something got lost in translation for me.

**Reply by pvvx** [reply]
Use it, who forbids it?

**Reply by evlo** [reply]
I do not understand how to set that up.

Only somewhat related thing i found might be this https://www.mysensors.org/ ?

I think I mainly do not understand how ble gateways in HA work.

**Reply by pvvx** [reply]
Sample: https://apps.rustore.ru/app/ru.kompius.thermo_plung
Supports LE long-range variant as well.

**Reply by evlo** [reply]
Oh ... the way is to use mqtt in the middle, did not occurred to me to to do it that way, thank you!

**Reply by pvvx** [reply]
Automatic translation (https://apps.rustore.ru/app/ru.kompius.thermo_plung): 
```
Designed for working with various wireless (BLE) temperature sensors, humidity, electric meters, etc.

It also allows you to connect to the data collectors (MQTT) for reading or transmitting measurements.

It is possible to connect 1 Wire sensors DS18x20, DS1990 (Ibutton) directly to the smartphone via OTG USB-Serial Converter (for example, CH340).

In the current version, the following sensors/devices are supported.

1. BLE GIGRO-thermo-rods of the production of NTC Compius LLC.
http://kompius.ru/

2. Xiaomi sensors both with the original firmware and optimized.
https://github.com/pvvx/atc_mithhermometer

3. Wire DS18x20

4. Electric meters CE308/208

Main functionality:

- search for devices in the area of the BLE SMARARTFON;

- reading data both from adversizing (advertising packages) and by installing the connection;

- data visualization: a list, on a grid, in the form of history trends;

- interaction with MQTT by a broker (reading, recording of readings);

- creation of user groups of devices;

-communication with data collectors (LLC "NTC Compius") for the formation of regular PDF reports sent to e-mail;

- Setting up variable sensors parameters.

Detailed documentation on the application page:
http://kompius.ru/products/48776765
```
The author does not provide source codes.
The author explains this by saying that there are a lot of examples of the Android API and writing your own application is very simple.
I have not communicated with other authors of similar software...
